Layoutlm onnx support

nishprabhu opened the pull request:
# What does this PR do?

This PR extends ONNX support to LayoutLM as explained in https://huggingface.co/transformers/serialization.html?highlight=onnx#converting-an-onnx-model-using-the-transformers-onnx-package

Fixes Issue  #13300


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@mfuntowicz @NielsRogge

nishprabhu committed:
Add support for exporting PyTorch LayoutLM to ONNX

nishprabhu committed:
Add support for exporting PyTorch LayoutLM to ONNX

nishprabhu committed:
Add support for exporting PyTorch LayoutLM to ONNX

nishprabhu committed:
Added tests for converting LayoutLM to ONNX

nishprabhu committed:
Added tests for converting LayoutLM to ONNX

nishprabhu committed:
Added tests for converting LayoutLM to ONNX

nishprabhu committed:
Merge branch 'huggingface:master' into layoutlm-onnx-support

nishprabhu committed:
cleanup

nishprabhu committed:
cleanup

nishprabhu committed:
Merge branch 'layoutlm-onnx-support' of https://github.com/nishprabhu/transformers into layoutlm-onnx-support

Cleanup

nishprabhu committed:
Removed regression/ folder

Narsil committed:
Fixing mbart50 with `return_tensors` argument too. (#13301)

* Fixing mbart50 with `return_tensors` argument too.

* Adding mbart50 tokenization tests.

patrickvonplaten committed:
[Flax] Correct all return tensors to numpy (#13307)

* fix_torch_device_generate_test

* remove @

* finish find and replace

anton-l committed:
Add Wav2Vec2 & Hubert ForSequenceClassification (#13153)

* Add hubert classifier + tests

* Add hubert classifier + tests

* Dummies for all classification tests

* Wav2Vec2 classifier + ER test

* Fix hubert integration tests

* Add hubert IC

* Pass tests for all classification tasks on Hubert

* Pass all tests + copies

* Move models to the SUPERB org

NielsRogge commented on src/transformers/models/layoutlm/configuration_layoutlm.py:
```diff
@@ -125,3 +131,84 @@ def __init__(
+        return OrderedDict(
+            [
+                ("input_ids", {0: "batch", 1: "sequence"}),
+                ("bbox", {0: "batch", 1: "sequence"}),
```
Oh how did you know that? Is this mentioned in the docs?

nishprabhu replied:
Its not explicitly mentioned in the docs. But the tutorial link you sent shows the config for Bert, and that does not specify the static dimension in the output (the 768 dimension in "last_hidden_state" and "pooler_output"). So I went through the code in src/transformers/onnx to verify this. The dimensions returned by the config file are passed on to the export function provided by PyTorch in the torch.onnx module as shown in src/transformers/onxx/conver.py:

```
    export(
        model,
        (model_inputs,),
        f=output.as_posix(),
        input_names=list(config.inputs.keys()),
        output_names=onnx_outputs,
        dynamic_axes={name: axes for name, axes in chain(config.inputs.items(), config.outputs.items())},
        do_constant_folding=True,
        use_external_data_format=config.use_external_data_format(model.num_parameters()),
        enable_onnx_checker=True,
        opset_version=opset,
    )

```
As you can see, the shapes provided by the config file are passed to the dynamic_axes parameter in the function call.

I also verified the change by converting a LayoutLM model to ONNX- the verification step that runs during conversion confirmed that the outputs of the converted model are similar to the original. Visualizing the converted ONNX model using Netron also shows that the input and output sizes are as follows:

```
Inputs:     
------     
name: input_ids    
type: int64[batch,sequence]    
    
name: bbox    
type: int64[batch,sequence,4]    
    
name: attention_mask    
type: int64[batch,sequence]    
    
name: token_type_ids    
type: int64[batch,sequence]    
    
Outputs:     
-------     
name: last_hidden_state    
type: float32[batch,sequence,768]    
    
name: pooler_output    
type: float32[batch,768]
```

I do think that this should be explicitly mentioned in the docs though, so that it will be easier for anyone trying to convert custom models.

mfuntowicz replied:
@nishprabhu you're totally right, this is about mapping the dynamic axis for each tensor (_i.e. axis we don't know beforehand or can change between two successive calls_).

In this case, if the bounding box dimensions are constant (x1, y2, x2, y2) then ONNX will infer and reports this in the traced model 👍🏻

mfuntowicz replied:
We will update the documentation to better reflect this 👍🏻

NielsRogge replied:
That's actually interesting to know, I've never worked with ONNX, but this seems indeed pretty important.

stefan-it committed:
examples: only use keep_linebreaks when reading TXT files (#13320)

* examples: only use keep_linebreaks when reading TXT files for all CLM examples

* examples: only use keep_linebreaks when reading TXT files for all CLM examples

* examples: only use keep_linebreaks when reading TXT files for all CLM examples

patrickvonplaten committed:
correct (#13304)

patrickvonplaten committed:
[Slow tests] Disable Wav2Vec2 pretraining test for now (#13303)

* fix_torch_device_generate_test

* remove @

* wav2vec2 pretraining

Co-authored-by: Patrick von Platen <patrick@huggingface.co>

Narsil committed:
Announcing the default model used by the pipeline (with a link). (#13276)

hwijeen committed:
use float 16 in causal mask and masked bias (#13194)

NielsRogge committed:
Add LayoutLMv2 + LayoutXLM (#12604)

* First commit

* Make style

* Fix dummy objects

* Add Detectron2 config

* Add LayoutLMv2 pooler

* More improvements, add documentation

* More improvements

* Add model tests

* Add clarification regarding image input

* Improve integration test

* Fix bug

* Fix another bug

* Fix another bug

* Fix another bug

* More improvements

* Make more tests pass

* Make more tests pass

* Improve integration test

* Remove gradient checkpointing and add head masking

* Add integration test

* Add LayoutLMv2ForSequenceClassification to the tests

* Add LayoutLMv2ForQuestionAnswering

* More improvements

* More improvements

* Small improvements

* Fix _LazyModule

* Fix fast tokenizer

* Move sync_batch_norm to a separate method

* Replace dummies by requires_backends

* Move calculation of visual bounding boxes to separate method + update README

* Add models to main init

* First draft

* More improvements

* More improvements

* More improvements

* More improvements

* More improvements

* Remove is_split_into_words

* More improvements

* Simply tesseract - no use of pandas anymore

* Add LayoutLMv2Processor

* Update is_pytesseract_available

* Fix bugs

* Improve feature extractor

* Fix bug

* Add print statement

* Add truncation of bounding boxes

* Add tests for LayoutLMv2FeatureExtractor and LayoutLMv2Tokenizer

* Improve tokenizer tests

* Make more tokenizer tests pass

* Make more tests pass, add integration tests

* Finish integration tests

* More improvements

* More improvements - update API of the tokenizer

* More improvements

* Remove support for VQA training

* Remove some files

* Improve feature extractor

* Improve documentation and one more tokenizer test

* Make quality and small docs improvements

* Add batched tests for LayoutLMv2Processor, remove fast tokenizer

* Add truncation of labels

* Apply suggestions from code review

* Improve processor tests

* Fix failing tests and add suggestion from code review

* Fix tokenizer test

* Add detectron2 CI job

* Simplify CI job

* Comment out non-detectron2 jobs and specify number of processes

* Add pip install torchvision

* Add durations to see which tests are slow

* Fix tokenizer test and make model tests smaller

* Frist draft

* Use setattr

* Possible fix

* Proposal with configuration

* First draft of fast tokenizer

* More improvements

* Enable fast tokenizer tests

* Make more tests pass

* Make more tests pass

* More improvements

* Addd padding to fast tokenizer

* Mkae more tests pass

* Make more tests pass

* Make all tests pass for fast tokenizer

* Make fast tokenizer support overflowing boxes and labels

* Add support for overflowing_labels to slow tokenizer

* Add support for fast tokenizer to the processor

* Update processor tests for both slow and fast tokenizers

* Add head models to model mappings

* Make style & quality

* Remove Detectron2 config file

* Add configurable option to label all subwords

* Fix test

* Skip visual segment embeddings in test

* Use ResNet-18 backbone in tests instead of ResNet-101

* Proposal

* Re-enable all jobs on CI

* Fix installation of tesseract

* Fix failing test

* Fix index table

* Add LayoutXLM doc page, first draft of code examples

* Improve documentation a lot

* Update expected boxes for Tesseract 4.0.0 beta

* Use offsets to create labels instead of checking if they start with ##

* Update expected boxes for Tesseract 4.1.1

* Fix conflict

* Make variable names cleaner, add docstring, add link to notebooks

* Revert "Fix conflict"

This reverts commit a9b46ce9afe47ebfcfe7b45e6a121d49e74ef2c5.

* Revert to make integration test pass

* Apply suggestions from @LysandreJik's review

* Address @patrickvonplaten's comments

* Remove fixtures DocVQA in favor of dataset on the hub

Co-authored-by: Lysandre <lysandre.debut@reseau.eseo.fr>

flaxel committed:
:sparkles: add citation file (#13214)

navjotts committed:
Improve documentation of pooler_output in ModelOutput (#13228)

* update documentation of pooler_output in modeling_outputs, making it more clear and available for generic usage

* Update src/transformers/modeling_outputs.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Update src/transformers/modeling_outputs.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* run make style

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

slowy07 committed:
fix: typo spelling grammar (#13212)

* fix: typo spelling grammar

* fix: make fixup

kamalkraj committed:
distilbert-flax (#13324)

* distilbert-flax

* added missing self

* docs fix

* removed tied kernal extra init

* updated docs

* x -> hidden states

* removed head_mask

* removed from_pt, +FLAX

* updated year

qqaatw committed:
Check None before going through iteration (#13250)

* Check None before going through iteration

* Format

sgugger committed:
Use existing functionality for #13251 (#13333)

fcakyon committed:
add ability to connect a neptune.ai run (#13319)

when `NEPTUNE_RUN_ID` environmetnt variable is set, neptune will log into the previous run with id `NEPTUNE_RUN_ID`

sgugger committed:
Update label2id in the model config for run_glue (#13334)

nateraw committed:
:bug: fix small model card bugs (#13310)

* :bug: fix small model card bugs

* :lipstick: style

mbforbes committed:
Fall back to `observed_batch_size` when the `dataloader` does not know the `batch_size`. (#13188)

bennimmo committed:
the use_auth_token has not been set up early enough in the model_kwargs. Fixes #12941 (#13205)

kamalkraj committed:
albert flax (#13294)

* albert flax

* year -> 2021

* docstring updated for flax

* removed head_mask

* removed from_pt

* removed passing attention_mask to embedding layer

qqaatw committed:
Correct wrong function signatures on the docs website (#13198)

* Correct outdated function signatures on website.

* Upgrade sphinx to 3.5.4 (latest 3.x)

* Test

* Test

* Test

* Test

* Test

* Test

* Revert unnecessary changes.

* Change sphinx version to 3.5.4"

* Test python 3.7.11

sgugger committed:
Fix AutoTokenizer when no fast tokenizer is available (#13336)

* Fix AutoTokenizer when a tokenizer has no fast version

* Add test

sgugger committed:
Fix release utils (#13337)

* Fix release utils

* Update docs/source/conf.py

Co-authored-by: Lysandre Debut <lysandre@huggingface.co>

Co-authored-by: Lysandre Debut <lysandre@huggingface.co>

laurahanu committed:
Add missing module __spec__ (#13321)

* added missing __spec__ to _LazyModule

* test __spec__ is not None after module import

* changed module_spec arg to be optional in _LazyModule

* fix style issue

* added module spec test to test_file_utils

tjruwase committed:
Use DS callable API to allow hf_scheduler + ds_optimizer (#13216)

* Use DS callable API to allow hf_scheduler + ds_optimizer

* Preserve backward-compatibility

* Restore backward compatibility

* Tweak arg positioning

* Tweak arg positioning

* bump the required version

* Undo indent

* Update src/transformers/trainer.py

* style

Co-authored-by: Stas Bekman <stas@stason.org>
Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

nishprabhu committed:
Fixed import error

nishprabhu committed:
Fixed merge conflicts in configuration_layoutlm.py

nishprabhu committed:
Remove unnecessary import statements

mfuntowicz commented:
@nishprabhu would you mind closing this PR and opening a new one, without any additional changes?

Github has a weird issue making the diff almost impossible to review.

Sorry for the inconvenience, please let us know if you need any help.

nishprabhu commented:
Sure, @mfuntowicz 
I'll open a new PR with the changes.
